File: frontend/src/types/chain.ts
/**
 * TypeScript types for Chain Orchestration.
 */

export enum ChainNodeType {
    AGENT = 'agent',
    CONDITION = 'condition',
    AGGREGATOR = 'aggregator',
    PARALLEL_SPLIT = 'parallel_split',
    PARALLEL_JOIN = 'parallel_join',
    START = 'start',
    END = 'end',
}

export enum ChainStatus {
    DRAFT = 'draft',
    ACTIVE = 'active',
    ARCHIVED = 'archived',
}

export enum ChainExecutionStatus {
    PENDING = 'pending',
    RUNNING = 'running',
    COMPLETED = 'completed',
    FAILED = 'failed',
    CANCELLED = 'cancelled',
    PAUSED = 'paused',
}

// ============================================================================
// Node & Edge Types for React Flow
// ============================================================================

export interface ChainNodeData {
    label: string
    agentId?: string
    agentName?: string
    nodeType: ChainNodeType
    config?: Record<string, any>
}

export interface ChainNode {
    id: string  // node_id
    type: string  // 'agentNode' | 'conditionNode' | etc. (React Flow node type)
    data: ChainNodeData
    position: { x: number; y: number }
}

export interface ChainEdge {
    id: string  // edge_id
    source: string  // source node_id
    target: string  // target node_id
    label?: string
    condition?: Record<string, any>
    type?: string  // Edge type for React Flow
}

// ============================================================================
// API Types (matching backend Pydantic schemas)
// ============================================================================

export interface ChainNodeSchema {
    node_id: string
    node_type: ChainNodeType
    agent_id?: string
    label: string
    position_x: number
    position_y: number
    config?: Record<string, any>
    order_index?: number
}

export interface ChainEdgeSchema {
    edge_id: string
    source_node_id: string
    target_node_id: string
    condition?: Record<string, any>
    label?: string
}

export interface ChainNodeResponse extends ChainNodeSchema {
    id: string
    chain_id: string
    created_at: string
    updated_at: string
}

export interface ChainEdgeResponse extends ChainEdgeSchema {
    id: string
    chain_id: string
    created_at: string
    updated_at: string
}

export interface Chain {
    id: string
    name: string
    description?: string
    status: ChainStatus
    version: string
    category?: string
    tags: string[]
    nodes: ChainNodeResponse[]
    edges: ChainEdgeResponse[]
    input_schema?: Record<string, any>
    output_schema?: Record<string, any>
    execution_count: number
    last_executed_at?: string
    created_at: string
    updated_at: string
    chain_metadata?: Record<string, any>
}

export interface ChainListItem {
    id: string
    name: string
    description?: string
    status: ChainStatus
    version: string
    category?: string
    tags: string[]
    node_count: number
    execution_count: number
    last_executed_at?: string
    created_at: string
    updated_at: string
}

export interface ChainCreateRequest {
    name: string
    description?: string
    category?: string
    tags?: string[]
    nodes: ChainNodeSchema[]
    edges: ChainEdgeSchema[]
    input_schema?: Record<string, any>
    output_schema?: Record<string, any>
    status?: ChainStatus
    metadata?: Record<string, any>
}

export interface ChainUpdateRequest {
    name?: string
    description?: string
    category?: string
    tags?: string[]
    nodes?: ChainNodeSchema[]
    edges?: ChainEdgeSchema[]
    input_schema?: Record<string, any>
    output_schema?: Record<string, any>
    status?: ChainStatus
    metadata?: Record<string, any>
}

export interface ChainExecuteRequest {
    input_data: Record<string, any>
    execution_name?: string
    variables?: Record<string, any>
    correlation_id?: string
}

export interface ChainExecution {
    id: string
    chain_id: string
    execution_name?: string
    status: ChainExecutionStatus
    input_data: Record<string, any>
    output_data?: Record<string, any>
    variables: Record<string, any>
    node_results: Record<string, any>
    started_at?: string
    completed_at?: string
    duration_seconds?: number
    current_node_id?: string
    completed_nodes: string[]
    active_edges?: string[]
    edge_results?: Record<string, any>
    error_message?: string
    error_details?: Record<string, any>
    correlation_id?: string
    created_at: string
    updated_at: string
}

export interface ChainExecutionListItem {
    id: string
    chain_id: string
    execution_name?: string
    status: ChainExecutionStatus
    started_at?: string
    completed_at?: string
    duration_seconds?: number
    error_message?: string
    created_at: string
}

export interface ChainExecutionLog {
    id: string
    execution_id: string
    node_id?: string
    event_type: string
    message: string
    level: string
    timestamp: string
    metadata?: Record<string, any>
}

export interface ChainValidationResult {
    is_valid: boolean
    errors: string[]
    warnings: string[]
    details: Record<string, any>
}

export interface ChainExecutionStatusResponse {
    execution_id: string
    status: ChainExecutionStatus
    current_node_id?: string
    active_edges?: string[]
    completed_nodes: string[]
    node_states?: Record<string, string>
    progress_percentage: number
    error_message?: string
}

// ============================================================================
// Utility Functions
// ============================================================================

/**
 * Convert backend ChainNodeResponse to React Flow node format.
 */
export function toReactFlowNode(node: ChainNodeResponse): ChainNode {
    return {
        id: node.node_id,
        type: getNodeTypeForReactFlow(node.node_type),
        data: {
            label: node.label,
            agentId: node.agent_id,
            nodeType: node.node_type,
            config: node.config || {},
        },
        position: {
            x: node.position_x,
            y: node.position_y,
        },
    }
}

/**
 * Convert backend ChainEdgeResponse to React Flow edge format.
 */
export function toReactFlowEdge(edge: ChainEdgeResponse): ChainEdge {
    return {
        id: edge.edge_id,
        source: edge.source_node_id,
        target: edge.target_node_id,
        label: edge.label,
        condition: edge.condition,
        type: 'smoothstep',  // or 'default', 'step', 'straight'
    }
}

/**
 * Convert React Flow node to backend ChainNodeSchema.
 */
export function toChainNodeSchema(node: ChainNode, index: number): ChainNodeSchema {
    return {
        node_id: node.id,
        node_type: node.data.nodeType,
        agent_id: node.data.agentId,
        label: node.data.label,
        position_x: node.position.x,
        position_y: node.position.y,
        config: node.data.config || {},
        order_index: index,
    }
}

/**
 * Convert React Flow edge to backend ChainEdgeSchema.
 */
export function toChainEdgeSchema(edge: ChainEdge): ChainEdgeSchema {
    return {
        edge_id: edge.id,
        source_node_id: edge.source,
        target_node_id: edge.target,
        condition: edge.condition,
        label: edge.label,
    }
}

/**
 * Get React Flow node type based on chain node type.
 */
function getNodeTypeForReactFlow(nodeType: ChainNodeType): string {
    switch (nodeType) {
        case ChainNodeType.AGENT:
            return 'agentNode'
        case ChainNodeType.CONDITION:
            return 'conditionNode'
        case ChainNodeType.AGGREGATOR:
            return 'aggregatorNode'
        case ChainNodeType.PARALLEL_SPLIT:
            return 'parallelSplitNode'
        case ChainNodeType.PARALLEL_JOIN:
            return 'parallelJoinNode'
        case ChainNodeType.START:
            return 'startNode'
        case ChainNodeType.END:
            return 'endNode'
        default:
            return 'default'
    }
}

// Type aliases for consistency
export type ChainNodeRequest = ChainNodeSchema
export type ChainEdgeRequest = ChainEdgeSchema

File: backend/shared/schemas/chain.py
"""Pydantic schemas for Chain Orchestration API."""

from typing import Optional, List, Dict, Any
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel, Field, ConfigDict

from shared.models.chain import ChainStatus, ChainNodeType, ChainExecutionStatus


# ============================================================================
# Node Schemas
# ============================================================================

class ChainNodeSchema(BaseModel):
    """Schema for a chain node."""
    node_id: str = Field(..., description="Unique identifier within the chain")
    node_type: ChainNodeType = Field(..., description="Type of node")
    agent_id: Optional[UUID] = Field(None, description="Agent ID if node type is AGENT")
    label: str = Field(..., description="Display label for the node")
    position_x: float = Field(0.0, description="X position on canvas")
    position_y: float = Field(0.0, description="Y position on canvas")
    config: Dict[str, Any] = Field(default_factory=dict, description="Node-specific configuration")
    order_index: int = Field(0, description="Execution order hint")

    model_config = ConfigDict(from_attributes=True)


class ChainNodeResponse(ChainNodeSchema):
    """Response schema for a chain node with database fields."""
    id: UUID
    chain_id: UUID
    created_at: datetime
    updated_at: datetime


# ============================================================================
# Edge Schemas
# ============================================================================

class ChainEdgeSchema(BaseModel):
    """Schema for a chain edge (connection between nodes)."""
    edge_id: str = Field(..., description="Unique identifier within the chain")
    source_node_id: str = Field(..., description="Source node ID")
    target_node_id: str = Field(..., description="Target node ID")
    condition: Optional[Dict[str, Any]] = Field(None, description="Conditional routing logic")
    label: Optional[str] = Field(None, description="Display label for the edge")

    model_config = ConfigDict(from_attributes=True)


class ChainEdgeResponse(ChainEdgeSchema):
    """Response schema for a chain edge with database fields."""
    id: UUID
    chain_id: UUID
    created_at: datetime
    updated_at: datetime


# ============================================================================
# Chain Schemas
# ============================================================================

class ChainCreateRequest(BaseModel):
    """Request schema for creating a new chain."""
    name: str = Field(..., min_length=1, max_length=255, description="Chain name")
    description: Optional[str] = Field(None, description="Chain description")
    category: Optional[str] = Field(None, description="Chain category")
    tags: List[str] = Field(default_factory=list, description="Tags for categorization")
    nodes: List[ChainNodeSchema] = Field(default_factory=list, description="Chain nodes")
    edges: List[ChainEdgeSchema] = Field(default_factory=list, description="Chain edges")
    input_schema: Optional[Dict[str, Any]] = Field(None, description="JSON schema for inputs")
    output_schema: Optional[Dict[str, Any]] = Field(None, description="JSON schema for outputs")
    status: ChainStatus = Field(ChainStatus.DRAFT, description="Chain status")
    metadata: Dict[str, Any] = Field(default_factory=dict, description="Additional metadata")


class ChainUpdateRequest(BaseModel):
    """Request schema for updating a chain."""
    name: Optional[str] = Field(None, min_length=1, max_length=255, description="Chain name")
    description: Optional[str] = Field(None, description="Chain description")
    category: Optional[str] = Field(None, description="Chain category")
    tags: Optional[List[str]] = Field(None, description="Tags for categorization")
    nodes: Optional[List[ChainNodeSchema]] = Field(None, description="Chain nodes")
    edges: Optional[List[ChainEdgeSchema]] = Field(None, description="Chain edges")
    input_schema: Optional[Dict[str, Any]] = Field(None, description="JSON schema for inputs")
    output_schema: Optional[Dict[str, Any]] = Field(None, description="JSON schema for outputs")
    status: Optional[ChainStatus] = Field(None, description="Chain status")
    metadata: Optional[Dict[str, Any]] = Field(None, description="Additional metadata")


class ChainResponse(BaseModel):
    """Response schema for a chain."""
    id: UUID
    name: str
    description: Optional[str]
    status: str
    version: str
    category: Optional[str]
    tags: List[str]
    nodes: List[ChainNodeResponse]
    edges: List[ChainEdgeResponse]
    input_schema: Optional[Dict[str, Any]]
    output_schema: Optional[Dict[str, Any]]
    execution_count: int
    last_executed_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime
    metadata: Dict[str, Any] = Field(default_factory=dict, alias="chain_metadata")

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class ChainListResponse(BaseModel):
    """Response schema for listing chains (lighter version)."""
    id: UUID
    name: str
    description: Optional[str]
    status: str
    version: str
    category: Optional[str]
    tags: List[str]
    node_count: int = Field(0, description="Number of nodes in the chain")
    execution_count: int
    last_executed_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ============================================================================
# Execution Schemas
# ============================================================================

class ChainExecuteRequest(BaseModel):
    """Request schema for executing a chain."""
    input_data: Dict[str, Any] = Field(..., description="Input data for the chain")
    execution_name: Optional[str] = Field(None, description="Name for this execution")
    variables: Optional[Dict[str, Any]] = Field(None, description="Initial variables")
    correlation_id: Optional[str] = Field(None, description="Correlation ID for tracking")
    model_override: Optional[Dict[str, Any]] = Field(None, description="Global model override for all agents in the chain")


class ChainExecutionResponse(BaseModel):
    """Response schema for chain execution."""
    id: UUID
    chain_id: UUID
    execution_name: Optional[str]
    status: str
    input_data: Dict[str, Any]
    output_data: Optional[Dict[str, Any]]
    variables: Dict[str, Any]
    node_results: Dict[str, Any]
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    duration_seconds: Optional[int]
    current_node_id: Optional[str]
    completed_nodes: List[str]
    active_edges: List[str] = Field(default_factory=list)
    edge_results: Optional[Dict[str, Any]] = None
    error_message: Optional[str]
    error_details: Optional[Dict[str, Any]]
    correlation_id: Optional[str]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ChainExecutionListResponse(BaseModel):
    """Response schema for listing executions (lighter version)."""
    id: UUID
    chain_id: UUID
    execution_name: Optional[str]
    status: str
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    duration_seconds: Optional[int]
    error_message: Optional[str]
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ChainExecutionLogResponse(BaseModel):
    """Response schema for execution logs."""
    id: UUID
    execution_id: UUID
    node_id: Optional[str]
    event_type: str
    message: str
    level: str
    timestamp: datetime
    metadata: Dict[str, Any] = Field(default_factory=dict, alias="log_metadata")

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


# ============================================================================
# Validation Schemas
# ============================================================================

class ChainValidationResult(BaseModel):
    """Result of chain validation."""
    is_valid: bool
    errors: List[str] = Field(default_factory=list)
    warnings: List[str] = Field(default_factory=list)
    details: Dict[str, Any] = Field(default_factory=dict)


class ChainExecutionStatusResponse(BaseModel):
    """Quick status response for execution."""
    execution_id: UUID
    status: ChainExecutionStatus
    current_node_id: Optional[str]
    completed_nodes: List[str]
    active_edges: List[str] = Field(default_factory=list)
    node_states: Optional[Dict[str, str]] = Field(None, description="Status of each node (pending, running, completed, failed, skipped)")
    progress_percentage: float = Field(0.0, ge=0.0, le=100.0)
    error_message: Optional[str]
